### ISSUE TITLE
simdjson_really_inline causes slow debug builds

### ISSUE THREAD

**Issue by strager** [simdjson/simdjson#1857]
**Describe the bug**
I have a project which uses simdjson. Debug (`-O0 -g -fsanitize=address,undefined`) builds with GCC 8 take a long time.

With simdjson version 1.0.0, a build of one file takes about 12 seconds.

If I change the `simdjson_really_inline` to be just `inline` (removing `__attribute__((always_inline))` from the definition), the build takes 2 seconds.

**To Reproduce**
(todo)

**simjson release**
1.0.0

**Configuration (please complete the following information if relevant)**
 - OS: Linux Mint, based on Ubuntu 20.04
 - Compiler: g++-8 (Ubuntu 8.4.0-3ubuntu2) 8.4.0
 - Optimization setting: -O0

**Indicate whether you are willing or able to provide a bug fix as a pull request**
Yes.

**Comment by strager** [comment]
For my purposes, allowing my project to customize `simdjson_really_inline` would work. Something like:

```cpp
#if !defined(simdjson_really_inline)
#define simdjson_really_inline inline __attribute__((always_inline))
#endif
```

Would a patch to do this be accepted?

**Comment by strager** [comment]
Perhaps another solution is to reduce the usage of `simdjson_really_inline`.

For example, I replaced `simdjson_really_inline` with `inline` for `operator[]` on `simdjson_result<SIMDJSON_IMPLEMENTATION::ondemand::value>`, and a debug build for that one file took 6 seconds (down from 11 seconds).

(With `simdjson_really_inline`, GCC generated over 30000 lines of assembly for just one call to `operator[]`. No wonder compilation takes a long time!)

**Comment by lemire** [comment]
Pull request invited.

**Comment by lemire** [comment]
At least with GCC and clang, we can detect 'optimized' built with

```
#ifdef __OPTIMIZE__
```

https://stackoverflow.com/questions/14618249/is-there-a-builtin-macro-defined-when-optimisation-is-enabled-in-clang


It might be wise to only trigger `__attribute__((always_inline))` when `__OPTIMIZE__` is defined.

**Comment by strager** [comment]
> It might be wise to only trigger `__attribute__((always_inline))` when `__OPTIMIZE__` is defined.

If the always_inline/__forceinline is conditional, should the macro be renamed? Perhaps to `simdjson_inline`.

**Comment by lemire** [comment]
Yep!

**Comment by lemire** [comment]
I think we should check for `NDEBUG` which controls whether assertions (in C) are checked. The benefit of it is that it is portable across compilers. It is part of the C and C++ standards.

**Comment by strager** [comment]
@lemire `NDEBUG` is often *not* defined in production. Debian discourages `NDEBUG`, and I think Debian projects using simdjson should not take a runtime performance hit.

**Comment by lemire** [comment]
@strager The simdjson library relies on `NDEBUG`. That's how you get runtime checks. If you push simdjson in production without `NDEBUG`, then you are going to get poor performance because of all the checks.

**Comment by lemire** [comment]
This is documented:

<img width="1189" alt="Screen Shot 2022-07-09 at 1 28 52 PM" src="https://user-images.githubusercontent.com/391987/178116576-626daa29-5e34-44cd-acbc-a393cddbeecc.png">


Of course, if you do not want to get high performance,  you may simply not set NDEBUG. 

I am not sure why one would use simdjson if high performance is not sought.

**Comment by strager** [comment]
Fixed in commit 5510089d4503ddcf70678bdc38d12b1fb5a1f6e5.
